### PR TITLE
fix getcandidate returns more delegates

### PR DESCRIPTION
**Pull request by lzxm160**

### CODE REVIEW

File: indexprotocol/blocks/protocol.go [hunk 1]
```diff
@@ -382,11 +385,17 @@ func (p *Protocol) updateDelegates(
 	if height >= p.epochCtx.FairbankHeight() {
 		return p.updateStakingDelegates(chainClient, height)
 	}
+	var preEpochStartHeight uint64
+	if epochNumber == 1 {
+		preEpochStartHeight = p.gravityChainCfg.GravityChainStartHeight
+	} else {
+		preEpochStartHeight = p.epochCtx.GetEpochHeight(epochNumber - 1)
+	}
```
Comment (lizhefeng) [inline indexprotocol/blocks/protocol.go:393]: if epochNumber == 1, we don't need to readState to get gravity chain height, just use gravityChainCfg.GravityChainStartHeight.

File: indexprotocol/rewards/protocol.go [hunk 2]
```diff
@@ -339,14 +342,21 @@ func (p *Protocol) updateCandidateRewardAddress(
...
 ) error {
 	if height >= p.epochCtx.FairbankHeight() {
 		return p.updateStakingCandidateRewardAddress(chainClient, height)
 	}
+	var preEpochStartHeight uint64
+	if epochNumber == 1 {
+		preEpochStartHeight = p.gravityChainCfg.GravityChainStartHeight
+	} else {
+		preEpochStartHeight = p.epochCtx.GetEpochHeight(epochNumber - 1)
+	}
```
Comment (lizhefeng) [inline indexprotocol/rewards/protocol.go:355]: ditto

File: indexprotocol/blocks/protocol.go [hunk 3]
```diff
@@ -382,11 +385,17 @@ func (p *Protocol) updateDelegates(
 	if height >= p.epochCtx.FairbankHeight() {
 		return p.updateStakingDelegates(chainClient, height)
 	}
+	var preEpochStartHeight uint64
+	if epochNumber == 1 {
+		preEpochStartHeight = p.gravityChainCfg.GravityChainStartHeight
```
Comment (lizhefeng) [inline indexprotocol/blocks/protocol.go:390]: preEpochStartHeight is epoch start height, which is different from gravityChainStartHeight.

File: indexprotocol/blocks/protocol.go [hunk 4]
```diff
@@ -375,43 +377,31 @@ func (p *Protocol) updateDelegates(
...
-			//retry to get chain start height again
-			return errors.New("waiting for fetching next timestamp in election service")
-		}
-		return nil
-	}, bo)
-	if nerr != nil {
-		return errors.Wrap(nerr, "failed to get gravity chain start height by backoff")
+	}
+	if gravityChainStartHeight == 0 {
+		//retry to get chain start height again
```
Comment (lizhefeng) [inline indexprotocol/blocks/protocol.go:398]: where is retry?

File: indexprotocol/rewards/protocol.go [hunk 5]
```diff
@@ -339,22 +342,24 @@ func (p *Protocol) updateCandidateRewardAddress(
...
+		gravityChainStartHeight, err = indexprotocol.GetGravityChainStartHeight(chainClient, prevEpochHeight)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get gravity height from chain service in epoch %d", epochNumber-1)
+		}
 	}
-	gravityChainStartHeight, err := strconv.ParseUint(string(readStateRes.GetData()), 10, 64)
-	if err != nil {
-		return errors.Wrap(err, "failed to parse gravityChainStartHeight")
+	if gravityChainStartHeight == 0 {
+		//retry to get chain start height again
```
Comment (lizhefeng) [inline indexprotocol/rewards/protocol.go:361]: No need to retry here.

**Comment by lizhefeng** [comment]
Pls fix tests

**Comment by lzxm160** [comment]
> Pls fix tests

sure

**Pull request merged**